### PR TITLE
[NUI] Add APIs for javascript popup.

### PR DESCRIPTION
**Pull request by huayongxu**
Support callbacks for 'url,changed', 'load.progress', and javascript
alert/confirm/prompt.

### API Changes ###

 - ACR:

### CODE REVIEW

File: src/Tizen.NUI/src/internal/Interop/Interop.WebView.cs [hunk 1]
```diff
@@ -105,6 +105,24 @@ internal static partial class WebView
             [global::System.Runtime.InteropServices.DllImport(NDalicPINVOKE.Lib, EntryPoint = "CSharp_Dali_WebView_AddJavaScriptMessageHandler")]
             public static extern void AddJavaScriptMessageHandler(global::System.Runtime.InteropServices.HandleRef jarg1, string jarg2, global::System.Runtime.InteropServices.HandleRef jarg3);
 
+            [global::System.Runtime.InteropServices.DllImport(NDalicPINVOKE.Lib, EntryPoint = "CSharp_Dali_WebView_RegisterJavaScriptAlertCallback")]
+            public static extern void WebView_RegisterJavaScriptAlertCallback(global::System.Runtime.InteropServices.HandleRef jarg1, global::System.Runtime.InteropServices.HandleRef jarg2);
```
Comment (huiyueun) [inline src/Tizen.NUI/src/internal/Interop/Interop.WebView.cs:109]: NUI use pascal case, could you delete 'WebView_'?

File: src/Tizen.NUI/src/internal/WebView/WebView.cs [hunk 2]
```diff
@@ -104,8 +117,11 @@ protected override void Dispose(DisposeTypes type)
                 //Release your own managed resources here.
                 //You should release all of your own disposable objects here.
                 pageLoadStartedSignal.Dispose();
+                pageLoadInProgressSignal.Dispose();
                 pageLoadFinishedSignal.Dispose();
                 pageLoadErrorSignal.Dispose();
+				scrollEdgeReachedSignal.Dispose();
```
Comment (huiyueun) [inline src/Tizen.NUI/src/internal/WebView/WebView.cs:123]: Please adjust the spacing. ;

File: src/Tizen.NUI/src/internal/WebView/WebView.cs [hunk 3]
```diff
@@ -134,6 +150,14 @@ private void OnPageLoadStarted(IntPtr data, string pageUrl)
             pageLoadStartedEventHandler?.Invoke(this, e);
         }
 
+        private void OnPageLoadInProgress(IntPtr data, string pageUrl)
+        {
+            WebViewPageLoadEventArgs e = new WebViewPageLoadEventArgs();
+            e.WebView = null; // uesless
+            e.PageUrl = null; // useless
+            pageLoadInProgressEventHandler?.Invoke(this, e);
```
Comment (huiyueun) [inline src/Tizen.NUI/src/internal/WebView/WebView.cs:158]: How about using this?
```csharp
pageLoadInProgressEventHandler?.Invoke(this, new WebViewPageLoadEventArgs());
```

File: src/Tizen.NUI/src/internal/WebView/WebView.cs [hunk 4]
```diff
@@ -161,6 +185,12 @@ private void OnScrollEdgeReached(IntPtr data, int edge)
             scrollEdgeReachedEventHandler?.Invoke(this, arg);
         }
 
+        private void OnUrlChanged(IntPtr data, string pageUrl)
+        {
+            WebViewUrlChangedEventArgs e = new WebViewUrlChangedEventArgs(pageUrl);
+            urlChangedEventHandler?.Invoke(this, e);
```
Comment (huiyueun) [inline src/Tizen.NUI/src/internal/WebView/WebView.cs:191]: How about using this?
```csharp
urlChangedEventHandler?.Invoke(this, new WebViewUrlChangedEventArgs(pageUrl));
```

File: src/Tizen.NUI/src/internal/WebView/WebViewUrlChangedEventArgs.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2019 Samsung Electronics Co., Ltd.
```
Comment (huiyueun) [inline src/Tizen.NUI/src/internal/WebView/WebViewUrlChangedEventArgs.cs:2]: Copyright (c) 2021 Samsung Electronics Co., Ltd.

File: src/Tizen.NUI/src/internal/WebView/WebViewUrlChangedEventArgs.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    /// Event arguments that passed via the WebView.PageLoadStarted or WebView.PageLoadFinished.
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public class WebViewUrlChangedEventArgs : EventArgs
+    {
+        /// <summary>
+        /// The url string of new web page.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Uri Url { get; }
```
Comment (huiyueun) [inline src/Tizen.NUI/src/internal/WebView/WebViewUrlChangedEventArgs.cs:33]: Is it correct to use the variable name 'url'?

**Review by zhouleonlei — APPROVED** [review]
LGTM

File: src/Tizen.NUI/src/internal/Interop/Interop.WebView.cs (hunk 1 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/internal/Interop/Interop.WebView.cs]: Done.

File: src/Tizen.NUI/src/internal/WebView/WebViewUrlChangedEventArgs.cs (hunk 6 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/internal/WebView/WebViewUrlChangedEventArgs.cs]: Renamed it to "NewPageUri".

File: src/Tizen.NUI/src/internal/WebView/WebViewUrlChangedEventArgs.cs (hunk 5 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/internal/WebView/WebViewUrlChangedEventArgs.cs]: Done

File: src/Tizen.NUI/src/internal/WebView/WebView.cs (hunk 4 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/internal/WebView/WebView.cs]: Done.

File: src/Tizen.NUI/src/internal/WebView/WebView.cs (hunk 3 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/internal/WebView/WebView.cs]: Done.

File: src/Tizen.NUI/src/internal/WebView/WebView.cs (hunk 2 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/internal/WebView/WebView.cs]: Done.

**Review by elishateng — APPROVED** [review]
LGTM

**Review by huiyueun — APPROVED** [review]
LGTM

**Comment by huiyueun** [comment]
@rabbitfor  Could you review this patch?

File: src/Tizen.NUI/src/internal/WebView/WebView.cs [hunk 7]
```diff
@@ -547,6 +573,32 @@ public bool VideoHoleEnabled
             }
         }
 
+        /// <summary>
+        /// Event for the PageLoadInProgress signal which can be used to subscribe or unsubscribe the event handler.<br />
+        /// This signal is emitted when page loading is in progress.<br />
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public event EventHandler<WebViewPageLoadEventArgs> PageLoadInProgress
```
Comment (rabbitfor) [inline src/Tizen.NUI/src/internal/WebView/WebView.cs:581]: [Event naming guideline for c# framework](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-events)

스크린샷 2021-02-26 오후 4 24 04

How about change name to `PageLoading`?

**Review by rabbitfor — COMMENTED** [review]
Could you following the ordering rule?
(I know it was already a mess, but it'd be great if you re-ordering it)

According to the StyleCop Rules Documentation the ordering is as follows.

Within a class, struct or interface: (SA1201 and SA1203)
* Constant Fields
* Fields
* Constructors
* Finalizers (Destructors)
* Delegates
* Events
* Enums
* Interfaces (interface implementations)
* Properties
* Indexers
* Methods
* Structs
* Classes

Within each of these groups order by access: (SA1202)
* public
* internal
* protected internal
* protected
* private

Within each of the access groups, order by static, then non-static: (SA1204)
* static
* non-static

File: src/Tizen.NUI/src/internal/WebView/WebView.cs (hunk 7 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/internal/WebView/WebView.cs]: Done

**Comment by huayongxu** [comment]
> * Fields

Thanks for your comments.
What does 'Fields' mean?
It means member variables of class?
If not, I will update this patch later.

**Comment by rabbitfor** [comment]
> > * Fields
> 
> Thanks for your comments.
> What does 'Fields' mean?
> It means member variables of class?
> If not, I will update this patch later.

Fields are variables!
Thank you.

**Pull request merged**